### PR TITLE
Design PBN Parser client interface

### PR DESCRIPTION
**Pull request by timheilman**
The goal is to design an API here for the consumer of a bridge PBN import, meaning, a description of a complete or incomplete bridge game's state.

These method calls will be the ones expected by the client to provide, so that the parser can call them.

If you were programming bridge-related things, what would make you happy or sad about this API?
 
@danielevans and/or @jharjung do you care to share this with other bridge-savvy programmers to try to solicit feedback on a good interface prior to coding to it?  (only with_dealt_card has been coded to so far).

### CODE REVIEW

File: spec/lib/portable_bridge_notation/shared_example_spec.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# For now, place to catch conception of full portable bridge notation
+# observer protocol
+
+# guarantee: each method here will be called at most once per
```
Comment (timheilman) [inline spec/lib/portable_bridge_notation/shared_example_spec.rb]: Oops, this guarantee isn't actually true, even for the example of with_dealt_card.  What's the best way to communicate as part of the protocol the promises regarding order and number of calls?

**Comment by timheilman** [comment]
In a Facebook chat with @jharjung, he advocates strongly against the event-driven/callback pattern here, on these bases: 

- the wonky parts of the promises regarding call order and count are a cognitive load on the client
- game boundaries aren't yet identified and even once they are, it's wrong to put the responsibility on the client to conceive of them

He prefers an enumerator pattern where the parser consumes the io and returns an enumerator for (or yields) a (big fat) PbnGame data structure, with a composition hierarchy containing data structures representing the results of the entire chatty event-driven pattern displayed here.

I felt the critiques are valid: I'm asking the client to keep track of a lot, and the call order and counts in the promises are completely wonky.  I don't believe using a data structure (and clients needing procedural, demeter-violating code to consume it) here would be bad: Uncle Bob says when it's more likely you'll need more functions on this data than that you'll need those functions to operate on new structures, go ahead with procedural over OO.

I'm still leery, though.  I hate nil, and feel it's a legitimate critique that with a standard as permissive for missing data as this one is, we could be passing back a data structure dripping with nils. Do I want to force the client to do *nil checks* on everything? No.  

Furthermore, the easiest thing to do to avoid feature envy on the structure is to wrap it in a domain object, which means I'm setting up the client to very likely depend directly not only on the interface for the parser's calls, but also on the format of the data structure that this parser chooses to return deep into the lifetime of the domain objects,  and without the client itself walking the tree (as this event-driven pattern is attempting to allow him to walk the PBN structure tree directly) that will be the case indefinitely.  If they're going to walk the tree anyway, why not walk the PBN tree directly via this event-driven pattern.  But they won't: the easiest (and thus likeliest) usage by client code would be long-lived composite hierarchies nil-laden like Christmas tree ornaments just lying in wait to throw an error. That doesn't sit well with me either.  Please, comments, anyone?

**Comment by timheilman** [comment]
Poking around on the web to see if this sort of dilemma has already been encountered and how others have solved it, I found this "prior art". It's parsing a different spec: XML, but with the same conundrum, which they solve by offering *both* event-driven *and* tree-based interaction: https://developer.apple.com/library/content/documentation/CoreFoundation/Conceptual/CFXML/Concepts/CFXMLParser.html

**Comment by jharjung** [comment]
Ok, that's an interesting analogy. So, when should one choose DOM parsing (tree-based) vs SAX parsing (event-driven)?

The disadvantage of DOM parsing is that it loads the entire document into memory. That's obviously a problem if the doc is too big to fit into memory. Even if you have gigabytes and gigabytes of RAM available, it's still less than ideal to parse a 4gb doc when you only need a few megabytes of that data.

SAX parsing was created to resolve that inefficiency. You can inspect the data as it's being loaded and choose to skip over big chunks of it, doing a "fast forward" through the IO stream until you hit the end of the tag you are skipping and then continuing from there. The downside is that dealing with SAX is more complicated than querying/manipulating a DOM tree.

Now we can ask, does PBN parsing have an analogous problem?

We have no upper limit on the size of the pbn files we might want to parse. So let's go ahead and assume that they could be gigantic.

In xml, any given subtree can be arbitrarily large. And the data model is general purpose - (nested) tags, attributes, etc. - which makes it difficult for the consuming client to declaratively specify which data it's interested in. Given those constraints, event-based parsing seems like a reasonable, if complex, solution.

These conditions don't hold for pbn though. The largest unit of data we have is a game, which will never be more than a couple kilobytes in size. And our data model is domain-specific. If skipping chunks of data becomes necessary for performance reasons, we could, with relative ease, implement configuration options to tell the parser which sections/tags/etc we are interested in.

Therefore, I would conclude that event-based parsing is not appropriate for PBN.

**Comment by timheilman** [comment]
Let me just say thanks so much for discussing this with me! I'm a firm believer that collaboration on difficult problems will generally result in better outcomes.

I agree that some structure is necessary here, and as the default if not exclusive interaction, because it eliminates client maintenance of parsing state, as you point out.  And one way to accomplish that without dealing with nil would be to make the field names of the structure themselves method names. Rather than nil checks, the client has respond_to? checks instead, or simply wing it and deal with problems raised when client code method invocations are missing on parts of the structure, probably by correcting the input stream.

**Comment by jharjung** [comment]
If I'm reading that correctly, the concern you are bringing up is how to deal with missing data. You are also, simultaneously, stating a dislike of nil. Let's try to untangle that.

The PBN spec allows for missing data, so we will need _some_ way to deal with that situation. It's unavoidable. So what options do we have? We can divide the possible approaches into two high level categories:

1) Model 'missing' as a value (nil, Option/Maybe, ...)
2) Model 'missing' as an error (raise, NoMethodError, ...)

Errors are intended for exceptional or unexpected situations. Since the spec allows for 'missing', it is by definition _not_ unexpected. So what possible justification do we have to model 'missing' as an error?

If we can't model missing as an error, then we must model it as a value. So how should we do that? 

1) Nil
This is ruby's native way to represent missing data. If we choose another representation, we _will_ find ourselves fighting against the language. That's not to say we must choose nil, but it should be a consideration. For example, if a user of our api wants to get rid of an existing value, they are likely to assume the way to do that is to set the field to nil.

2) Option/Maybe
This is how functional languages do missing, and it definitely has some perks. There are implementations of this for ruby too. Here are a couple I found with a cursory google search:
https://github.com/bhb/maybe
https://github.com/rap1ds/ruby-possibly
(Just throwing those out there for illustration. There may be better implementations available.)

3) Something else?

**Pull request merged**